fcrepo/fcrepo
# add resource locking

**JIRA Ticket**: https://jira.lyrasis.org/browse/FCREPO-3123

**This PR requires https://github.com/fcrepo/fcrepo-storage-ocfl/pull/25 to work**

# What does this Pull Request do?

Adds resource locking so that a resource cannot be concurrently modified in different transaction.

In this implementation,

* A transaction holds a lock on a resource for the duration of the transaction
* The same transaction may modify a resource that it holds a lock on multiple times
* AGs are locked when any of their children are modified
* Binaries are locked when their description is modified
* Binary descriptions are locked when their binary is modified

# How should this be tested?

1. Create a resource
2. Start a transaction
3. Modify the resource in a transaction, but do not commit
4. Attempt to modify the same resource outside of the transaction. This should fail with a `409`.

# Interested parties
@fcrepo/committers
